[5.8] Add deleted_at date cast for soft deleting models

timacdonald opened the pull request:
This PR utilises a trait initializer to automatically add `deleted_at` to the `$dates` array of any model that uses the `SoftDeletes` trait. Previously this has to be added manually to each model - unlike the `created_at` and `updated_at` that is handled "under the hood".

In essence...

```php
class Post extends Eloquent
{
    use SoftDeletes;

    // no need for this anymore
    // protected $dates = ['deleted_at'];
}
```

**Notes:**
- An existing`$casts` entry for `deleted_at` will take precedence (test exists).
- A duplicate `$dates` entry for `deleted_at` will have no impact as the array is not a key => value pair, but a value array. Also when `getDates` is called it already runs `array_unique` to remove any duplicate entries (test exists).
- An existing mutator `getDeletedAtAttribute` will take precedence (test exists).

**Breaking:**

This will be a breaking change if you are not expecting the value of `deleted_at` to be cast, i.e. you never added it to the `$dates` array as [outlined in the docs](https://laravel.com/docs/5.7/eloquent#deleting-models). But as mentioned earlier, a mutator will take precedence if for whatever reason you don't want the automatic casting.

timacdonald committed:
add deleted_at date cast for soft deleted models

taylorotwell commented:
Is there any path for opting out of the cast?

timacdonald commented:
Just added a new test to confirm the current behaviour can be maintained by adding a cast.

This is the easiest opt-out and the one I see being in the upgrade guide if this PR was to be accepted. A `boolean` toggle, as used for `created_at` and `updated_at`, seemed like overkill as the cast is a one-liner.

```php
protected $casts = [
    'deleted_at' => 'string', // retrain current functionality
]; 
```

It is also possible to opt-out with an attribute mutator - but obviously that is more than a one liner. 

If you think a `boolean` toggle would be better for consistency I am happy to add that instead.

timacdonald commented:
I actually think I was mistaken and the cast will cover all cases as a opt-out. If you prefer a `boolean` flag approach I can add the following to this PR...

```php
trait SoftDeletes
{
    protected $castDeletedAtToDateInstance = true;

    public function initializeSoftDeletes()
    {
        if ($this->castDeletedAtToDateInstance) {
            $this->dates[] = $this->getDeletedAtColumn();
        }
    }
    
    // ...
}
```

And then the dev can opt-out with...

```php
class Post extends Eloquent
{
    use SoftDeletes;

    protected $castDeletedAtToDateInstance = false;
}
```

timacdonald commented:
I think the boolean flag is the better approach after sitting on this for the day so I've implemented that, but in a seperate commit so you can roll it back if you prefer the previous approach.

driesvints commented:
Not that I have a better alternative but the current property name looks super confusing and long..

taylorotwell commented:
Eh, I think I prefer the string cast to the property.

timacdonald commented:
Just reverted and removed the boolean flag. However Travis is said to have failed - but as I'm looking at it right now it actually hasn't run the tests. Not sure if you can whack it with a big spanner to get it to run again?

taylorotwell commented:
Merged.
